### PR TITLE
build 2.12.1 :snowflake:

### PR DESCRIPTION
**Pull request by boldorider4**
emoji 2.12.1 :snowflake:

**Destination channel:** Snowflake

### Links

- [PKG-4900]
- [Upstream repository](https://github.com/carpedm20/emoji/tree/v2.12.1)
- [conda-forge-feedstock](https://github.com/AnacondaRecipes/emoji-feedstock/blob/main/recipe/meta.yaml)    

### Explanation of changes:
- patched tests because of missing `testutils`

[PKG-4900]: https://anaconda.atlassian.net/browse/PKG-4900?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,39 +1,67 @@
...
   host:
-    - python >=2.7
+    - python
     - pip
+    - setuptools
+    - wheel
   run:
-    - python >=2.7
+    - python
+    - typing-extensions
```
Comment (ViridianMelody) [inline recipe/meta.yaml]:
```suggestion
    - typing-extensions >=4.7.0
```

**Review by ViridianMelody — APPROVED** [review]
Looks good aside from the missing `typing-extensions` version pinning

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,39 +1,67 @@
...
-  sha256: a2986c21e4aba6b9870df40ef487a17be863cb7778dcf1c01e25917b7cd210bb
+  sha256: 4aa0488817691aa58d83764b6c209f8a27c0b3ab3f89d1b8dceca1a62e4973eb
+  patches:
+    - 0001-factor-out-testutils.patch
 
 build:
-  noarch: python
   number: 0
-  script: {{ PYTHON }} -m pip install . --no-deps -vv
+  script: {{ PYTHON }} -m pip install . --no-deps --no-build-isolation -vv
```
Comment (JamesRobertsonGames) [inline recipe/meta.yaml]: Requires python version bound
```suggestion
        skip: True  # [py<37]
```

https://github.com/carpedm20/emoji/blob/v2.12.1/pyproject.toml#L9C1-L9C26

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,39 +1,67 @@
...
 
 requirements:
+  build:
+    - patch     # [not win]
+    - m2-patch  # [win]
   host:
-    - python >=2.7
+    - python
     - pip
+    - setuptools
```
Comment (JamesRobertsonGames) [inline recipe/meta.yaml:25]: Require bound
```suggestion
    - setuptools >=65.0.0
```

https://github.com/carpedm20/emoji/blob/v2.12.1/pyproject.toml#L2C24-L2C32

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,39 +1,67 @@
...
+{% set tests_to_skip = tests_to_skip + ' or test_get_emoji_by_alias' %}
 
 test:
   imports:
     - emoji
+  source_files:
+    - tests
+  requires:
+    - pip
+    - pytest
```
Comment (JamesRobertsonGames) [inline recipe/meta.yaml]: Requires bound
```suggestion
    - pytest >=7.4.4
```

https://github.com/carpedm20/emoji/blob/v2.12.1/pyproject.toml#L46

**Review by skupr-anaconda — APPROVED** [review]

File: recipe/meta.yaml (hunk 3 above)
Comment (boldorider4) [reply recipe/meta.yaml:25]: I suppose at this point it is a style choice whether we want to preserve all upstream requirements, even though they are not applicable to our build process. I think we go back and forth with this. We have not observed requirements for modern `setuptools` in a while.

**Pull request merged**